adc/ade7913/ade7913-driver.c

```diff
@@ -293,256 +448,292 @@ static int edma_configure(void)
+static int edma_configure(void)
 {
-	int res;
-	oid_t dir;
-	msg_t msg;
+	int devnum, res, i;
+	oid_t *oid = OID_NULL;
 
-	res = portCreate(&common.port);
-	if (res != EOK) {
-		log_error("could not create port: %d", res);
-		return -1;
+	res = mutexCreate(&common.edma_spi_rcv_lock);
+	if (res < 0) {
+		log_error("Mutex resource creation failed");
+		return res;
 	}
 
-	res = mkdir(ADC_DEVICE_DIR, 0);
-	if (res < 0 && errno != EEXIST) {
-		log_error("mkdir /dev failed (%d)", -errno);
-		return -1;
+	res = condCreate(&common.edma_spi_rcv_cond);
+	if (res < 0) {
+		log_error("Conditional resource creation failed");
+		return res;
 	}
 
-	if ((res = lookup(ADC_DEVICE_DIR, NULL, &dir)) < 0) {
-		log_error("%s lookup failed (%d)", ADC_DEVICE_DIR, res);
-		return -1;
+	res = condCreate(&common.dready_cond);
+	if (res < 0) {
+		log_error("Conditional resource creation failed");
+		return res;
 	}
 
-	msg.type = mtCreate;
-	msg.i.create.type = otDev;
-	msg.i.create.mode = 0;
-	msg.i.create.dev.port = common.port;
-	msg.i.create.dev.id = 0;
-	msg.i.create.dir = dir;
-	msg.i.data = ADC_DEVICE_FILE_NAME;
-	msg.i.size = sizeof(ADC_DEVICE_FILE_NAME);
-	msg.o.data = NULL;
-	msg.o.size = 0;
+	common.buff = mmap(NULL, (ADC_BUFFER_SIZE + _PAGE_SIZE - 1) / _PAGE_SIZE * _PAGE_SIZE,
+		PROT_READ | PROT_WRITE, MAP_UNCACHED, oid, common.buffer_paddr);
 
-	if ((res = msgSend(dir.port, &msg)) < 0 || msg.o.create.err != EOK) {
-		log_error("could not create %s (res=%d, err=%d)",
-			ADC_DEVICE_FILE_NAME, res, msg.o.create.err);
-		return -1;
+	if (common.buff == MAP_FAILED) {
+		log_error("eDMA buffer allocation failed");
+		return -ENOMEM;
 	}
 
-	log_info("device initialized");
+	memset(common.buff, 0, ADC_BUFFER_SIZE);
+	common.buffer_paddr = va2pa(common.buff);
 
-	return 0;
-}
+	/* Set request commands order */
+	for (i = 0; i < common.devcnt; ++i) {
+		devnum = (int)(common.order[i] - '0');
+		spi_write_cmd_lookup[i] = (spi_write_cmd_lookup[i] & ~(0x3000000)) | (devnum << 24);
+	}
 
+	res = dma_setup_tcds();
+	if (res < 0) {
+		log_error("Failed to init TCDs");
+		return res;
+	}
 
-static int dev_open(oid_t *oid, int flags)
-{
-	(void)oid;
-	(void)flags;
+	interrupt(EDMA_CHANNEL_IRQ(SPI_RCV_DMA_CHANNEL),
+		edma_spi_rcv_irq_handler, NULL, common.edma_spi_rcv_cond, &common.edma_spi_ch_handle);
+
+	dma_start();
 
 	return EOK;
 }
 
-static int dev_close(oid_t *oid, int flags)
+
+static int dev_init(const char *devname)
 {
-	(void)oid;
-	(void)flags;
+	int res = portCreate(&common.oid.port);
+	if (res != EOK) {
+		log_error("Could not create port: %d", res);
+		return res;
+	}
+
+	res = create_dev(&common.oid, devname);
+	if (res < 0) {
+		log_error("Could not create %s (res=%d)", devname, res);
+		return res;
+	}
+
+	log_info("Device initialized");
 
 	return EOK;
 }
 
+
 static int dev_read(void *data, size_t size)
 {
 	int res = 0;
 
-	if (data != NULL && size != sizeof(unsigned))
+	if (data != NULL && size != sizeof(unsigned)) {
 		return -EIO;
+	}
```

Comment:
What will be in line `675` if `data == NULL`?
I think condition should be like this:
```suggestion
	if (data == NULL || size != sizeof(unsigned)) {
		return -EIO;
	}
```

Reply:
You're right, thanks for noticing this.